crates/shared/src/app_data.rs

```diff
@@ -0,0 +1,103 @@
+use {
+    anyhow::{anyhow, Context, Result},
+    model::app_id::AppDataHash,
+    serde::Deserialize,
+    serde_json::Value,
+};
+
+#[derive(Debug)]
+pub struct ValidatedAppData {
+    pub hash: AppDataHash,
+    pub backend: BackendAppData,
+}
+
+#[derive(Debug, Default, Deserialize)]
+pub struct BackendAppData {}
+
+#[derive(Clone)]
+pub struct Validator {
+    pub size_limit: usize,
+}
+
+impl Default for Validator {
+    fn default() -> Self {
+        Self { size_limit: 1000 }
+    }
+}
+
+impl Validator {
+    pub fn validate(&self, full_app_data: &[u8]) -> Result<ValidatedAppData> {
+        // Empty full app data is allowed even though it isn't valid json. It matches
+        // all zero app data (no hashing performed). Makes life easier for API
+        // consumers, who don't care about app data and offers some degree of backward
+        // compatibility.
+        if full_app_data.is_empty() {
+            return Ok(ValidatedAppData {
+                hash: AppDataHash([0u8; 32]),
+                backend: Default::default(),
+            });
+        }
+
+        if full_app_data.len() > self.size_limit {
+            return Err(anyhow!(
+                "app data has size {} which is larger than limit {}",
+                full_app_data.len(),
+                self.size_limit
+            ));
```

Comment:
nit: Should we specify in which format the message is supposed to be smaller than the size limit?
For example the user might escape all the `"` or unicode characters which would mean the raw string they send us could be way bigger than the data we actually store on disk.
I think just looking at the raw bytes of the sent string is fine (anything else would probably be a lot work for little improvement) but making the error message more explicit would be low hanging fruit.

Reply:
I'll add this somewhere. Either this message or the openapi docs.